lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,350 @@
+
+        int fd = shm_open(shm_name.c_str(), (O_CREAT | O_RDWR), (S_IRUSR | S_IWUSR));
+
+        if (fd == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Cannot open shared memory named {:s}: {:s}"), shm_name, strerror(errno)));
+        }
+
+        // Resize object to hold buffer
+        if (ftruncate(fd, shm_size) == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to expand shared memory named {:s}: {:s}"), shm_name, strerror(errno)));
+        }
+        INFO("Resized to {} bytes\n", (long) shm_size);
+
+        addr = (uint8_t*) mmap(nullptr, shm_size, (PROT_READ | PROT_WRITE), MAP_SHARED, fd, 0);
+        if (addr == MAP_FAILED) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to map shm {:s} to memory: {:s}."), shm_name, strerror(errno)));
+        }
+
+        // fd is no longer needed
+        if (close(fd) == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to close file descriptor for shm {:s}: {:s}."), shm_name, strerror(errno)));
+        }
+
+        return addr;
+}
+
+bool visSharedMemWriter::add_sample(const visFrameView& frame, time_ctype t, uint32_t freq_ind) {
+    // process the time sample
+
+    if ( vis_time_ind_map.count(t) != 0) {
+        // if the time is already indexed, write to memory at that location
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+    else if (vis_time_ind_map.size() < ntime) {
+        // if there is still empty space in the shared buffer
+        // add an additional time index
+        cur_pos++;
+        vis_time_ind_map[t] = cur_pos;
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+
+    time_ctype max_time = vis_time_ind_map.rbegin()->first;
+    time_ctype min_time = vis_time_ind_map.begin()->first;
+
+    if (t < min_time) {
+        // this data is older than anything else in the map, so we should
+        // just drop it
+        INFO("Dropping integration as buffer (FPGA count: {:d}) arrived too late (minimum in pool {:d})", t.fpga_count, min_time.fpga_count);
+        return false;
+    }
+
+    else if (t > max_time) {
+        // we need to drop the oldest time
+        // and replace it with the new time index
+        cur_pos++;
+        reset_memory(cur_pos);
+        vis_time_ind_map.erase(min_time);
+        vis_time_ind_map[t] = cur_pos;
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+    else {
+        INFO("Dropping integration as buffer (FPGA count: {:d}) arrived too late (only accepting new times greater than {:d})", t.fpga_count, max_time.fpga_count);
+        return false;
+    }
+}
+
+void visSharedMemWriter::reset_memory(uint32_t time_ind) {
+
+    uint8_t *buf_write_pos = buf_addr + (time_ind * nfreq * frame_size);
+    uint64_t *record_write_pos = record_addr + time_ind;
```

Comment:
This needs to be changed to reflect a 2 dimensional access record (per time, per frequency)